File: Data Structures & Algorithms/three-integer-sum/submission-6.py
"""
Input: 
    - nums : List[Int]
Output:
    - List[Int] : list of triplets that sum up to 0
Constraints:
    - output should not contain duplicates
    - 3 <= nums.length <= 1000
    - -10^5 <= nums[i] <= 10^5

Plan: Sort the nums list in acsending order and go through each element in the 
list. Then use two sum on the window of elements occuring after the current 
element to find two numbers that can add up to the current element to be 0. 

[-4, -1, -1, 0, 1, 2]

Psuedo Code:
sort the nums list
res = []

for i in range(len(nums)):
    If the current element is a duplicate, continue until the last occurence

    left = i + 1
    right = len(nums)-1

    while left < right:
        currentSum = nums[i] + nums[left] + nums[right]

        if currentSum is equal to 0:
            add the triple to the list

            update the left and right pointer

            update the left pointer to skip duplicates
            while nums[left] == nums[left-1]:
                left += 1

        otherwise, if the currentSum > 0:
            decrement the right pointer
        else:
            increment the left pointer 

return result
"""

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        res = []

        for i in range(len(nums)):
            if i != 0 and nums[i] == nums[i-1]:
                continue

            left = i + 1
            right = len(nums) - 1

            while left < right:
                currentSum = nums[i] + nums[left] + nums[right]

                if currentSum == 0:
                    res.append([nums[i], nums[left], nums[right]])

                    left += 1
                    right -= 1

                    while nums[left] == nums[left-1] and left < right:
                        left += 1
                elif currentSum > 0:
                    right -= 1
                else:
                    left += 1

        return res


        

File: Data Structures & Algorithms/max-water-container/submission-6.cs
/*
Input:
    - int[] : heights
Output:
    - int : max amoiunt of water a container can store
Constraints:
    - length of heights: [2,100000]
    - values of heights: [0,10000]

length = rightIndex - leftIndex
height = min(rightIndex, leftIndex)
Area = lengh * height

Plan: Two pointers
1. Create variables
    - maxArea = 0 
    - left = 0 
    - right = len(heights) - 1
2. while left < right:
        length = right - left
        height = min(right, left)
        area = length * height

        maxArea = max(maxArea, area)

        if leftIndex > rightIndex:
            decrement right
        else
            increment left 
3. Return maxArea
*/

public class Solution {
    public int MaxArea(int[] heights) {
        int maxArea = 0;
        int left = 0;
        int right = heights.Length - 1;

        while (left < right) { 
            int area = (right - left) * Math.Min(heights[left], heights[right]);
            maxArea = Math.Max(maxArea, area);

            // update pointers
            if (heights[left] > heights[right]) {
                right--;
            }
            else {
                left++;
            }
        }

        return maxArea;
    }
}
